### PR TITLE
Modernize map notes UI

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

 ```SUMMARY: Interface "Modernize map notes UI with `uilist`"```

#### Purpose of change

Modernize map notes UI.

#### Describe the solution

- added keybinding to toggle display of map notes;
- utilized `uilist` to display of map notes.

#### Additional context

I've used uilist-callbacks, so it is possible to edit and delete notes directly from the list. There are also all kind of useful info - total number of notes in current overmap buffer, distance from each note to player, note coordinates, location description. Oh - live note preview is also supported when editing or moving through the list.

If your current overmap tile contains a note, uilist will automatically switch to corresponding row.

__Screenshot__:

image

__Video__:

https://imgur.com/a/nwhkScp

### CODE REVIEW

**Comment by AMurkin** [comment]
Typos in the OP:
```diff
-```SUMMARY: Interface "Modernize map notes UI with uilist"
+```SUMMARY: Interface "Modernize map notes UI with uilist"```

-Modernizeв map notes UI.
+Modernized map notes UI.
```

It may be useful to show the direction from the current overmap using the arrows (as we show the direction of the wind).

File: src/overmap_ui.cpp [hunk 1]
```diff
@@ -290,86 +290,38 @@ static void draw_camp_labels( const catacurses::window &w, const tripoint &cente
...
+        const std::string note_symbol = std::string( 1, std::get<0>( om_symbol ) );
+        const std::string note_text = note.substr( std::get<2>( om_symbol ), std::string::npos );
+        nmenu.addentry_desc( string_format( "[%s] %s", colorize( note_symbol, note_color ),
+                                            note_text ),
+                             string_format( "( <color_white>%d</color>, <color_white>%d</color> ) %s", p.x, p.y,
+                                            overmap_buffer.get_description_at( origin, true ) ) );
+        row++;
+    }
+    nmenu.query();
+    if( nmenu.ret >= 0 && nmenu.ret < notes.size() ) {
```
Comment (ZhilkinSerg) [inline src/overmap_ui.cpp]:
```suggestion
    if( nmenu.ret >= 0 && nmenu.ret < static_cast<int>( notes.size() ) ) {
```

File: src/overmap_ui.cpp [hunk 2]
```diff
@@ -290,86 +292,120 @@ static void draw_camp_labels( const catacurses::window &w, const tripoint &cente
...
+        refresh = false;
+        nmenu.init();
+        g->refresh_all();
+        nmenu.desc_enabled = true;
+        nmenu.input_category = "OVERMAP_NOTES";
+        nmenu.additional_actions.emplace_back( "DELETE_NOTE", "" );
+        nmenu.additional_actions.emplace_back( "EDIT_NOTE", "" );
+        const input_context ctxt( nmenu.input_category );
+        nmenu.text = string_format(
+                         _( "<%s> - show note, <%s> - edit note, <%s> - delete note, <%s> - close window" ),
```
Comment (ZhilkinSerg) [inline src/overmap_ui.cpp]:
```suggestion
                         _( "<%s> - center on note, <%s> - edit note, <%s> - delete note, <%s> - close window" ),
```

**Comment by ZhilkinSerg** [comment]
Did you try using color tags in notes? E.g.:

<color_red>blablabla</color>

**Comment by ZhilkinSerg** [comment]
An option to toggle coloring note text based on note symbol color could be easily implemented.

When this option is enabled, function `remove_color_tags` need to be used on note text to remove all color tags which can possible exist in the string, then `colorize` should be used with note symbol color on note text with original color tags removed.

**Comment by Treah** [comment]
> Did you try using color tags in notes? E.g.:
> 
> <color_red>blablabla
Ohh that totally works! Good enough for me :+1:

**Pull request merged**